### PR TITLE
Feature/Test helper for awaiting the output of Promise

### PR DESCRIPTION
**Pull request by EgzonArifi**
- PromiseKit unit test helper in order to test async Promise base code
- covered with tests

### CODE REVIEW

File: Tests/Tests/PromiseKit/XCTestCaseTests+Promise.swift [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  func test_awaitPromise_returnsFailureResultWithExpectedError() throws {
+    let anyError = anyNSError()
+    
+    let failurePromise: Promise<TestDouble> = anyFailurePromise(anyError)
+    switch try awaitPromise(failurePromise) {
+    case .failure(let error as NSError):
+      XCTAssertEqual(anyError as NSError, error)
+    case .success:
+      XCTFail("Expected failure result")
+    }
```
Comment (gentique) [inline Tests/Tests/PromiseKit/XCTestCaseTests+Promise.swift:33]: We could refactor this with a helper method `expect`

```swift
  func expect<T: Equatable>(_ promise: Promise<T>, toCompleteWith expectedResult: Result<T, Error>, file: StaticString = #file, line: UInt = #line) throws {
    switch try (awaitPromise(promise), expectedResult) {
    case let (.success(success), .success(expectedSuccess)):
      XCTAssertEqual(success, expectedSuccess, "Expected \(expectedSuccess) got \(success) instead", file: file, line: line)
    case let (.failure(error as NSError), .failure(expectedError as NSError)):
      XCTAssertEqual(error, expectedError, "Expected \(expectedError) got \(error) instead", file: file, line: line)
    default:
      XCTFail("Expected \(expectedResult)", file: file, line: line)
    }
  }
```
Then this test could look like: 
```swift
    let anyError = anyNSError()
    let failurePromise: Promise<TestDouble> = anyFailurePromise(anyError)
    
    try expect(failurePromise, toCompleteWith: .failure(anyError))
```

The `test_awaitPromise_returnsSuccessResultWithExpectedObject` test would look like: 
```swift
    let anyObject = anyObject()
    let successPromise = anySuccessPromise(anyObject)
    
    try expect(successPromise, toCompleteWith: .success(anyObject))
```

**Review by gentique — APPROVED** [review]
I like it 👌

**Review by yllfejziu — APPROVED** [review]
👌

**Review by markom-povio — APPROVED** [review]

**Review by borut-t — APPROVED** [review]
I like the idea but am not sure how much it brings to the table. Anyway, I don't have strong feelings for or against it. Let's merge it 👌

**Comment by EgzonArifi** [comment]
> I like the idea but am not sure how much it brings to the table. Anyway, I don't have strong feelings for or against it. Let's merge it 👌

if you take a look into the tests, you can see how it simplifies the testing of Promise-base code. So now on your scenarios, you can instantly  assert your Promise Output without having to setup an `expectation`, wait and full fill it every time you test a Promise function.

**Pull request merged**